### DISCUSSION TITLE
Can we no longer use Mysql views in the new update 1.0.2?

### DISCUSSION THREAD

**Discussion by KernOil** [Budibase/budibase#3650] [Help]
Before I was able to use views to write and view specific data. is that not available anymore? I pull in the id_key from the main table and that worked. how can I continue using my views?

![image](https://user-images.githubusercontent.com/80058238/144462862-17357f83-2eca-4af8-81b6-c79043acf690.png)

**Comment by mjashanks** [comment]
What has happened here is that we've put in better error reporting for fetching tables - we had an issue with tables with no primary keys.

Don't think that we realised that "Fetch Tables" was returning views.

Although how did this work for writing? Are views not read-only? Aren't they essentially just stored `select` queries?

**Reply by KernOil** [reply]
No Idea. I too thought views were just for reading. however, a while back I added the main tables row id column and I could write to my views in BB like it was a table. was really cool. I used all views for my subform table and row submission edits.
![image](https://user-images.githubusercontent.com/80058238/144482423-181c21a6-c033-4cac-b9ef-128b00b9d7ac.png)
May not have been best practice but worked perfectly for months! 
Do you know the direction you plan to go with views? will they be supported?

**Reply by KernOil** [reply]
it looks like writing to a view is actually a thing. using this format...

INSERT INTO [table_name]
       SELECT [row number],         --this is why adding the tables row id did the trick
              [data],

**Reply by mjashanks** [reply]
Honestly, I don't think we should be fetching views at all in here (it was probably unintentional). Is there a reason why you can't just use the table? What is the view doing for you?

You can still of course use views via custom queries - `select ... from myView` - but then use the actual table for insert, update, delete.

**Reply by KernOil** [reply]
I was using Views because of the way the DB is set up. it's a flat-file instead of a star schema. ill just need to redesign the DB to a relational DB.

**Reply by daraul** [reply]
What if budibase could be told to treat a particular column as an index? I have some table components that are served fairly well by "saved queries" in BB, but then I lose pagination on those tables (https://github.com/Budibase/budibase/discussions/7313).

Or, have the table scroll infinitely.

**Comment by amrtdc1** [comment]
@mjashanks, I am also able to query views in my MS SQL database strictly for reading, but I have a different, but potentially related issue now.  The new structure for establishing a connection to my database times out.  I'm assuming this is because of the vast number of tables in the database and BB is attempting to pull all of them in and read them.  Thus, I can't even make the data connection needed to build queries for my data source.  Is it possible to utilize the Fetch Tables button, but also keep the option for building queries on the main screen when configuring your data source?  Otherwise, I'm not sure I'll be able to establish a connection again due to the timeout error. Thanks!

**Reply by amrtdc1** [reply]
@mjashanks , I was able to establish a connection to the master database on the same server, and then change the database and generate queries, so there is a way to make it work if you can establish a connection to a smaller database first using the "Fetch Tables" button, and then go back into the data source config and change the database you are targeting. Thanks!

**Reply by mjashanks** [reply]
Thanks Drew - we should add a "Skip" button, so we don't have to fetch tables. But thank you for reporting & telling us the workaround.

**Reply by UnusualPi** [reply]
@mjashanks it looks like it's now performing a primary key validation which is why views aren't showing, but it will also prevent tables without primary keys to show.  This is a pretty critical need for displaying some types of information and keeping indexes small when not needed.  I'm guessing the validation on the primary key constraint was introduced to simplify the cud parts of crud in the latest update.  Maybe you could still display the tables but only allow read on them.

**Reply by mjashanks** [reply]
Hi @UnusualPi - you are correct - we need the primary key for the CUD parts to work.

Technically, we don't need a primary key for `SELECT many` or even `INSERT`.  However, we would also need some way for Budibase to understand that tables without a PK only have these capabilities. This would touch a lot of areas in Budibase.

It's possible to add a custom query of course, which seems like a reasonable workaround - unless you have a lot of tables that this applies to.

We're open to covering this scenario in the future. However, I can't see it being a priority for us in the near term given that it looks like an edge case, with a sensible workaround. 

Happy for you to prove me wrong here - does your database have a lot of tables without a primary key? Do you need to query a lot of these in Budibase?

**Reply by UnusualPi** [reply]
Hi @mjashanks makes sense and I don't have an unreasonable amount to work within my test environment.  But, if I am looking to scale out to my user base I anticipate a rather large influx of tickets coming into my data ops team.  So, it would be a nice to have.  Appreciate the quick response!
